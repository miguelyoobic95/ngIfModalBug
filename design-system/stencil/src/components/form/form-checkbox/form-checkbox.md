---
name: Checkbox
category: Form Components
---

## Attributes

|Name|Type|Default|Description|
|---|---|---|---|
|`text`|string|   |the text of the checkbox|
|`state`|string| unchecked |the state of the checkbox, can be one of three values; unchecked, indeternimate and checked|
|`disabled`|boolean| `false`  |whether the checkbox is enabled or not|
|`isIndeterminate`|boolean| | allows a third indeterminate state for the checkbox|

```yoo-form-checkbox.html
<div style="padding:10px;">
    <yoo-form-checkbox text="this is a checkbox" class="accent"></yoo-form-checkbox>
    <br><br>
    <yoo-form-checkbox text="this is an indeterminate checkbox" class="accent" is-indeterminate="true"></yoo-form-checkbox>
    <br><br>
    <yoo-form-checkbox text="this is a checkbox with starting state checked" state="checked" class="accent"></yoo-form-checkbox>
    <br><br>
    <yoo-form-checkbox text="this is a disabled unchecked checkbox" disabled="true"></yoo-form-checkbox>
    <br><br>
    <yoo-form-checkbox text="this is a disabled indeterminate checkbox" disabled="true" state="indeterminate"></yoo-form-checkbox>
    <br><br>
    <yoo-form-checkbox text="this is a disabled checked checkbox" disabled="true" state="checked"></yoo-form-checkbox>
</div>
```

## CSS

The checkbox has the `dark` class by default.

|Type|Name|Description|
|---|---|---|
|Large|`large`|yoo-form-checkbox large|
|Accent|`accent`|yoo-form-checkbox accent|
|Success|`success`|yoo-form-checkbox success|
|Danger|`danger`|yoo-form-checkbox danger|
|Info|`info`|yoo-form-checkbox info|
|Warning|`warning`|yoo-form-checkbox warning|
|Gradient Accent|`gradient-accent`|yoo-form-checkbox gradient-accent|
|Gradient Dark|`gradient-dark`|yoo-form-checkbox gradient-dark|
|Gradient Success|`gradient-success`|yoo-form-checkbox gradient-success|
|Gradient Danger|`gradient-danger`|yoo-form-checkbox gradient-danger|
|Gradient Info|`gradient-info`|yoo-form-checkbox gradient-info|
|Gradient Warning|`gradient-warning`|yoo-form-checkbox gradient-warning|

```yoo-form-checkbox-styles.html
<div style="padding:10px;">
    <yoo-form-checkbox text="this is a large accent checkbox" state="checked" class="accent large"></yoo-form-checkbox>
    <br><br>
    <yoo-form-checkbox text="this is an accent checkbox" class="accent" state="checked"></yoo-form-checkbox>
    <br><br>
    <yoo-form-checkbox text="this is a danger checkbox" class="danger" state="checked"></yoo-form-checkbox>
    <br><br>
    <yoo-form-checkbox text="this is a success checkbox" class="success" state="checked"></yoo-form-checkbox>
    <br><br>
    <yoo-form-checkbox text="this is an info checkbox" class="info" state="checked"></yoo-form-checkbox>
    <br><br>
    <yoo-form-checkbox text="this is a warning checkbox" class="warning" state="checked"></yoo-form-checkbox>
    <br><br>
    <yoo-form-checkbox text="this is a dark checkbox (default)" class="dark" state="checked"></yoo-form-checkbox>
</div>
<div style="padding:10px;">
    <yoo-form-checkbox text="this is a large accent checkbox" state="checked" class="round accent large"></yoo-form-checkbox>
    <br><br>
    <yoo-form-checkbox text="this is an accent checkbox" class="round accent" state="checked"></yoo-form-checkbox>
    <br><br>
    <yoo-form-checkbox text="this is a danger checkbox" class="round danger" state="checked"></yoo-form-checkbox>
    <br><br>
    <yoo-form-checkbox text="this is a success checkbox" class="round success" state="checked"></yoo-form-checkbox>
    <br><br>
    <yoo-form-checkbox text="this is an info checkbox" class="round info" state="checked"></yoo-form-checkbox>
    <br><br>
    <yoo-form-checkbox text="this is a warning checkbox" class="round warning" state="checked"></yoo-form-checkbox>
    <br><br>
    <yoo-form-checkbox text="this is a dark checkbox (default)" class="round dark" state="checked"></yoo-form-checkbox>
</div>
```

## Events

|Attr|Description|
|---|---|---|
|`onCheck`|event emitted when a checkbox state is changed|