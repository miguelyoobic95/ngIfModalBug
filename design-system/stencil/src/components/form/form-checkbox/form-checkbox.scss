
@mixin checkbox-color($class, $color) {
    :host(.#{$class}) {
        .container {
            .icon-container {
                 &.enabled {
                    background-color: cssvar($color);
                    border-color: cssvar($color);
                }
            }
        }
    }
}

:host {
    .container {
        display: inline-flex;
        align-items: center;
        .icon-container {
            transition: all 0.3s;
            border-radius: 5px;
            min-width: 1.25rem;
            height: 1.25rem;
            border: 0.1rem solid;
            justify-content: center;
            align-items: center;
            &.enabled {
                background-color: cssvar(dark-40);
                border-color: cssvar(dark-40);
                    &:hover{
                        cursor: pointer;
                        }
                }
            &.disabled {
                background-color: cssvar(dark-20);
                border-color: cssvar(dark-20);
            }
            &.empty {
                background-color: cssvar(light);
                &.enabled {
                    background-color: cssvar(light);
                    &:hover{
                        cursor: pointer;
                        }
                }
                &.disabled {
                    background-color: cssvar(dark-20);
                    border-color: cssvar(dark-20);
                }
            }
                .icon {
                    color: cssvar(light);
                    transition: all 0.3s;
                    font-size: 0.875rem;
                    padding-top: 0.125rem;
                    &.disabled {
                        color: cssvar(dark-80);
                    }
                    &.empty {
                        opacity: 0;
                    }
                }
        }    
        .text-container {
            padding-left: 0.5rem;
            display: flex;
            justify-content: flex-start;
            color: cssvar(dark);
            &.disabled {
                color: cssvar(dark-20);
            }
            &.enabled {
                 &:hover{
                cursor: pointer;
            }
            }   
        }
    }
}

:host(.round) {
    .container {
        .icon-container {
            border-radius: 50%;
        }
    }
}

:host(.large) {
    .container {
        .icon-container {
            width: 2rem;
            height: 2rem;
            .icon {
                font-size: 1rem;
                padding-top: 0rem;
            }
        }
    }
}

@include checkbox-color(accent, accent);
@include checkbox-color(danger, danger);
@include checkbox-color(success, success);
@include checkbox-color(info, info);
@include checkbox-color(warning, warning);
