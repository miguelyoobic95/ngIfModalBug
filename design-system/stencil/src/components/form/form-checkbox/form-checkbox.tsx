import { Component, Prop, Event, EventEmitter, Method, Element } from '@stencil/core';

@Component({
    tag: 'yoo-form-checkbox',
    styleUrl: 'form-checkbox.scss',
    scoped: true
})
export class YooFormCheckboxComponent {

    @Prop() text: string;
    @Prop({ mutable: true }) state: string = 'unchecked';
    @Prop() disabled: boolean;
    @Prop() isIndeterminate: boolean;

    @Event() checkboxToggled: EventEmitter<string>;

    @Element() host: HTMLStencilElement;

    @Method()
    onCheckboxClick() {
        this.getNextState();
        this.checkboxToggled.emit(this.state);
    }

    getNextState() {
        const TRANSITIONS = {
            checked: 'unchecked',
            indeterminate: 'checked',
            unchecked: this.isIndeterminate ? 'indeterminate' : 'checked'
        };
        this.state = TRANSITIONS[this.state];
    }

    render(): JSX.Element {
        return (
            <div class="container">
                {this.disabled ?
                    <div class={this.state === 'unchecked' ? 'icon-container empty disabled' : 'icon-container disabled'} attr-layout="row">
                        <span class={this.state === 'unchecked' ? 'icon empty' :  'icon disabled'}><i class={this.state === 'indeterminate' ? 'yo-minus' : 'yo-check-solid'}></i></span>
                    </div>
                    :
                    <div class={this.state === 'unchecked' ? 'icon-container empty enabled' : 'icon-container enabled'} attr-layout="row" onClick={() => this.onCheckboxClick()}>
                        <span class={this.state === 'unchecked' ? 'icon empty' :  'icon'}><i class={this.state === 'indeterminate' ? 'yo-minus' : 'yo-check-solid'}></i></span>
                    </div>
                }
                {this.disabled ?
                    <div class="text-container disabled">
                        {this.text}
                    </div> :
                    <div class="text-container enabled" onClick={() => this.onCheckboxClick()}>
                        {this.text}
                    </div>
                }
            </div>
        );
    }
}