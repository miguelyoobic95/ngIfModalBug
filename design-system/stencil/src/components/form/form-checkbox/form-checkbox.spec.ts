import { TestWindow } from '@stencil/core/testing';
import { YooFormCheckboxComponent } from './form-checkbox';

describe('YooFormCheckboxComponent', () => {
    it('should build', () => {
        expect(new YooFormCheckboxComponent()).toBeTruthy();
    });

    describe('Snapshots', () => {
        it ('should have text and class', async () => {
            let window = new TestWindow();
            let element = await window.load({
                components: [YooFormCheckboxComponent],
                html: '<yoo-form-checkbox text="this is text" class="accent"></yoo-form-checkbox>'
            });
            expect(element).toMatchSnapshot();
        });

        it ('should have indeterminate', async () => {
            let window = new TestWindow();
            let element = await window.load({
                components: [YooFormCheckboxComponent],
                html: '<yoo-form-checkbox text="this is text" class="accent" indeterminate="true"></yoo-form-checkbox>'
            });
            expect(element).toMatchSnapshot();
        });

         it ('should have checkced state', async () => {
             let window = new TestWindow();
             let element = await window.load({
                components: [YooFormCheckboxComponent],
                html: '<yoo-form-checkbox text="this is text" class="accent" state="checked"></yoo-form-checkbox>'
            });
            expect(element).toMatchSnapshot();
        });

         it ('should be disabled', async () => {
             let window = new TestWindow();
             let element = await window.load({
                components: [YooFormCheckboxComponent],
                html: '<yoo-form-checkbox text="this is text" disabled="true"></yoo-form-checkbox>'
            });
            expect(element).toMatchSnapshot();
        });
    });

    describe('rendering', () => {
       it('should re-render when state is updated', async () => {
           let window = new TestWindow();
           let element = await window.load({
               components: [YooFormCheckboxComponent],
               html: '<yoo-form-checkbox text="this is text" class="accent"></yoo-form-checkbox>'
           });
           expect(element.state).toEqual('unchecked');
           element.state = 'checked';
           await window.flush();

           expect(element.state).toEqual('checked');
       });

       it('should render state and text', async () => {
           let window = new TestWindow();
           let element = await window.load({
               components: [YooFormCheckboxComponent],
               html: '<yoo-form-checkbox text="this is text" state="checked" class="accent"></yoo-form-checkbox>'
           });
           expect(element.state).toEqual('checked');
           expect(element.textContent).toEqual('this is text');
       });
    });

    describe('events', () => {
        it ('should emit on container click', async () => {
            let window = new TestWindow();
            let element = await window.load({
                components: [YooFormCheckboxComponent],
                html: '<yoo-form-checkbox></yoo-form-checkbox>'
            });

            let res;
            let emitItem = jest.fn(ev => { res = ev.detail; });
            window.document.addEventListener('checkboxToggled', emitItem);
            await element.querySelector('.icon-container').click();
            expect(emitItem).toHaveBeenCalled();
            expect(res).toEqual('checked');
        });

        it ('should emit unchecked on container click two times', async () => {
            let window = new TestWindow();
            let element = await window.load({
                components: [YooFormCheckboxComponent],
                html: '<yoo-form-checkbox></yoo-form-checkbox>'
            });

            let res;
            let emitItem = jest.fn(ev => { res = ev.detail; });
            window.document.addEventListener('checkboxToggled', emitItem);
            await element.querySelector('.icon-container').click();
            await element.querySelector('.text-container').click();
            expect(emitItem).toHaveBeenCalled();
            expect(res).toEqual('unchecked');
        });
    });
});
