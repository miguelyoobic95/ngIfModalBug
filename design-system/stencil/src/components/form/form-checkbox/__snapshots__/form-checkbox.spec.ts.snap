// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`YooFormCheckboxComponent Snapshots should be disabled 1`] = `
<yoo-form-checkbox
  class="hydrated"
  data-yoo-form-checkbox-host=""
  disabled="true"
  text="this is text"
>
  
  <div
    class="container"
    data-yoo-form-checkbox=""
  >
    <div
      attr-layout="row"
      class="icon-container empty disabled"
      data-yoo-form-checkbox=""
    >
      <span
        class="icon empty"
        data-yoo-form-checkbox=""
      >
        <i
          class="yo-check-solid"
          data-yoo-form-checkbox=""
        />
      </span>
    </div>
    <div
      class="text-container disabled"
      data-yoo-form-checkbox=""
    >
      this is text
    </div>
  </div>
</yoo-form-checkbox>
`;

exports[`YooFormCheckboxComponent Snapshots should have checkced state 1`] = `
<yoo-form-checkbox
  class="accent hydrated"
  data-yoo-form-checkbox-host=""
  state="checked"
  text="this is text"
>
  
  <div
    class="container"
    data-yoo-form-checkbox=""
  >
    <div
      attr-layout="row"
      class="icon-container enabled"
      data-yoo-form-checkbox=""
    >
      <span
        class="icon"
        data-yoo-form-checkbox=""
      >
        <i
          class="yo-check-solid"
          data-yoo-form-checkbox=""
        />
      </span>
    </div>
    <div
      class="text-container enabled"
      data-yoo-form-checkbox=""
    >
      this is text
    </div>
  </div>
</yoo-form-checkbox>
`;

exports[`YooFormCheckboxComponent Snapshots should have indeterminate 1`] = `
<yoo-form-checkbox
  class="accent hydrated"
  data-yoo-form-checkbox-host=""
  indeterminate="true"
  text="this is text"
>
  
  <div
    class="container"
    data-yoo-form-checkbox=""
  >
    <div
      attr-layout="row"
      class="icon-container empty enabled"
      data-yoo-form-checkbox=""
    >
      <span
        class="icon empty"
        data-yoo-form-checkbox=""
      >
        <i
          class="yo-check-solid"
          data-yoo-form-checkbox=""
        />
      </span>
    </div>
    <div
      class="text-container enabled"
      data-yoo-form-checkbox=""
    >
      this is text
    </div>
  </div>
</yoo-form-checkbox>
`;

exports[`YooFormCheckboxComponent Snapshots should have text and class 1`] = `
<yoo-form-checkbox
  class="accent hydrated"
  data-yoo-form-checkbox-host=""
  text="this is text"
>
  
  <div
    class="container"
    data-yoo-form-checkbox=""
  >
    <div
      attr-layout="row"
      class="icon-container empty enabled"
      data-yoo-form-checkbox=""
    >
      <span
        class="icon empty"
        data-yoo-form-checkbox=""
      >
        <i
          class="yo-check-solid"
          data-yoo-form-checkbox=""
        />
      </span>
    </div>
    <div
      class="text-container enabled"
      data-yoo-form-checkbox=""
    >
      this is text
    </div>
  </div>
</yoo-form-checkbox>
`;
